src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.Crypto.cs

```diff
@@ -170,5 +171,106 @@ internal static byte[] GetDynamicBuffer<THandle>(NegativeSizeReadMethod<THandle>
+        private static unsafe partial int SetMemoryTracking(delegate* unmanaged<MemoryOperation, UIntPtr, UIntPtr, int, char*, int, void> trackingCallback);
+
+        [StructLayout(LayoutKind.Sequential)]
+        private unsafe struct MemoryEntry
+        {
+            public int Size;
+            public int Line;
+            public char* File;
+        }
+
+        private enum MemoryOperation
+        {
+            Malloc = 1,
+            Realloc = 2,
+            Free = 3,
+        }
+
+        private static readonly unsafe nuint Offset = (nuint)sizeof(MemoryEntry);
+        // We only need to store the keys but we use ConcurrentDictionary to avoid locking
+        private static ConcurrentDictionary<UIntPtr, UIntPtr>? _allocations;
+
+        [UnmanagedCallersOnly]
+        private static unsafe void MemoryTrackinCallback(MemoryOperation operation, UIntPtr ptr, UIntPtr oldPtr, int size, char* file, int line)
+        {
+            ref MemoryEntry entry = ref *(MemoryEntry*)ptr;
+
+            Debug.Assert(entry.File != null);
+            Debug.Assert(ptr != UIntPtr.Zero);
+
+            switch (operation)
+            {
+                case MemoryOperation.Malloc:
+                    Debug.Assert(size == entry.Size);
+                    _allocations!.TryAdd(ptr, ptr);
+                    break;
+                case MemoryOperation.Realloc:
+                    if ((IntPtr)oldPtr != IntPtr.Zero)
+                    {
+                        _allocations!.TryRemove(oldPtr, out _);
+                    }
+                    _allocations!.TryAdd(ptr, ptr);
+                    break;
+                case MemoryOperation.Free:
+                    _allocations!.TryRemove(ptr, out _);
+                    break;
+            }
+        }
+
+        public static unsafe void EnableTracking()
+        {
+            _allocations ??= new ConcurrentDictionary<UIntPtr, UIntPtr>();
+            _allocations!.Clear();
+            SetMemoryTracking(&MemoryTrackinCallback);
+        }
+
+        public static unsafe void DisableTracking()
+        {
+            SetMemoryTracking(null);
+            _allocations!.Clear();
+        }
+
+        public static unsafe Tuple<UIntPtr, int, string>[] GetIncrementalAllocations()
+        {
+            if (_allocations == null || _allocations.IsEmpty)
+            {
+                return Array.Empty<Tuple<UIntPtr, int, string>>();
+            }
+
+            Tuple<UIntPtr, int, string>[] allocations = new Tuple<UIntPtr, int, string>[_allocations.Count];
+            int index = 0;
+            foreach ((UIntPtr ptr, UIntPtr value) in _allocations)
+            {
+                ref MemoryEntry entry = ref *(MemoryEntry*)ptr;
+                allocations[index] = new Tuple<UIntPtr, int, string>(ptr + Offset, entry.Size, $"{Marshal.PtrToStringAnsi((IntPtr)entry.File)}:{entry.Line}");
+                index++;
+            }
```

Comment:
Can the `_allocations.Count` value change while iterating over `_allocations`?

Reply:
yes. I expect the `ConcurrentDictionary` to handle parallel access gracefully but you may not get consistent snapshot. We did it while back with the `HashSet` and manual locking. Since the iteration can take a while I feel it is better to allow concurrent access and deliver what we can. In ideal situation one would enable tracking run the repro and dump whatever it remains.